internal/controller/nimpipeline_controller.go

```diff
@@ -33,30 +39,257 @@ type NIMPipelineReconciler struct {
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the NIMPipeline object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.2/pkg/reconcile
 func (r *NIMPipelineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	// Fetch the NIMPipeline instance
+	nimPipeline := &appsv1alpha1.NIMPipeline{}
+	if err := r.Get(ctx, req.NamespacedName, nimPipeline); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			logger.Error(err, "unable to fetch NIMPipeline", "name", req.NamespacedName)
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	logger.Info("Reconciling", "NIMPipeline", nimPipeline.Name)
+
+	// Check if the instance is marked for deletion
+	if nimPipeline.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Add finalizer if not present
+		if !controllerutil.ContainsFinalizer(nimPipeline, NIMPipelineFinalizer) {
+			controllerutil.AddFinalizer(nimPipeline, NIMPipelineFinalizer)
+			if err := r.Update(ctx, nimPipeline); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		// The instance is being deleted
+		if controllerutil.ContainsFinalizer(nimPipeline, NIMPipelineFinalizer) {
+			// Perform cleanup of resources
+			if err := r.cleanupNIMPipeline(ctx, nimPipeline); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			// Remove finalizer to allow for deletion
+			controllerutil.RemoveFinalizer(nimPipeline, NIMPipelineFinalizer)
+			if err := r.Update(ctx, nimPipeline); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, nil
+		}
+	}
 
-	// TODO(user): your logic here
+	// Handle nim-cache reconciliation
+	if result, err := r.reconcileNIMPipeline(ctx, nimPipeline); err != nil {
+		logger.Error(err, "error reconciling NIMCache", "name", nimPipeline.Name)
+		return result, err
+	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *NIMPipelineReconciler) cleanupNIMPipeline(ctx context.Context, nimPipeline *appsv1alpha1.NIMPipeline) error {
+	// TODO: Add custom cleanup
+
+	// All owned objects are garbage collected
+	return nil
+}
+
+func (r *NIMPipelineReconciler) reconcileNIMPipeline(ctx context.Context, nimPipeline *appsv1alpha1.NIMPipeline) (ctrl.Result, error) {
+	// Track enabled NIMServices
+	enabledServices := make(map[string]bool)
+
+	// Process each service specification in the pipeline
+	for _, service := range nimPipeline.Spec.Services {
+		enabled := service.Enabled != nil && *service.Enabled
+		enabledServices[service.Name] = enabled
+
+		if !enabled {
+			continue
+		}
+
+		if err := r.reconcileNIMService(ctx, nimPipeline, service); err != nil {
```

Comment:
Update the condition of NIMPipeline obj here so it does not fail silently.

Reply:
Changed the logic to continue reconciliation of all services before erroring out. We log the error and capture the status after all services are reconciled. Conditions would need to be added as a separate PR.